compiler/rustc_middle/src/ty/instance.rs

```diff
@@ -426,7 +426,10 @@ impl<'tcx> fmt::Display for Instance<'tcx> {
 }
 
 impl<'tcx> Instance<'tcx> {
-    pub fn new(def_id: DefId, args: GenericArgsRef<'tcx>) -> Instance<'tcx> {
+    /// Creates a new [`InstanceKind::Item`] from the `def_id` and `args`. Note that this
+    /// does not resolve trait items to their corresponding impl items -- to do that, use
+    /// [`Instance::expect_resolve`] instead.
```

Comment:
As a compiler dev I would be rather clueless about whether I want to "resolve trait items to their corresponding impl items" or not -- I'm not even sure what exactly this means. So I'm afraid this comment isn't very helpful in terms of giving guidance to when I should use which of these functions.

Reply:
Do you have a suggestion for a new comment? Because just saying it is not very helpful as a comment is... well, also not very actionable 😆

Reply:
See the comment on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Instance.html#method.try_resolve to understand what it means to resolve a trait item to its corresponding impl item.

Reply:
> Do you have a suggestion for a new comment? Because just saying it is not very helpful as a comment is... well, also not very actionable 😆

I'm not sure how to suggest a better comment without having any clue about when this function should and should not be used.^^ You made const-eval use `new_raw` so I guess it is the right thing to use for anything I encountered in the compiler so far...

Reply:
> You made const-eval use new_raw so I guess it is the right thing to use for anything I encountered in the compiler so far...

CTFE uses `Instance::expect_resolve` all throughout, so that's definitely not true. The only usage sites I've touched are the ones that attempt to directly evaluate const bodies.

Reply:
> I guess what matters is: "Are you certain that the body of `def_id` is the actual body you want to use. This is not the case for trait methods or functions we may stub out with a separate implementation"

That sounds helpful. Are there docs for the concept of "resolve" that we can link to? I guess it refers to going from a trait-method pair to the actual impl? Also, is there ever a downside to using `resolve` when we don't have to?

`new_raw` is scary as a name so now I feel like every use of this function needs a comment justifying it, but I'm not sure what the justification looks like in the interpreter.